### PR TITLE
[ysh] implement len() as vm.Callable

### PR DESCRIPTION
**Pull request by melvinw**
This patch implements `len()` over `value_t`. It also fixes some error presentation code in the command evaluator

### CODE REVIEW

File: library/func_misc.py [hunk 1]
```diff
@@ -147,3 +147,44 @@ def Call(self, pos_args, named_args):
...
+        # type: () -> None
+        pass
+
+    def Call(self, pos_args, named_args):
+        # type: (List[value_t], Dict[str, value_t]) -> value_t
+
+        n_args = len(pos_args)
+        if n_args != 1:
+            raise error.InvalidType("len() expects exactly one argument but %d were given" %
+                          n_args, loc.Missing)
```
Comment (melvinw) [inline library/func_misc.py:164]: Following the conventions in `frontend/typed_args.py` for consistency. I noticed python also raises `TypeError` in this case, but it feels like it should be something else...

**Comment by melvinw** [comment]
Nice. Healthy jump in `delta-cpp` for the ysh spec tests from this. Down to 116 w/ 420/536 passing. Was 141 before

http://travis-ci.oilshell.org/github-jobs/4753/cpp-spec.wwz/_tmp/spec/ysh-cpp/compare.html

File: library/func_misc.py [hunk 2]
```diff
@@ -147,3 +147,44 @@ def Call(self, pos_args, named_args):
...
+
+    def __init__(self):
+        # type: () -> None
+        pass
+
+    def Call(self, pos_args, named_args):
+        # type: (List[value_t], Dict[str, value_t]) -> value_t
+
+        n_args = len(pos_args)
+        if n_args != 1:
```
Comment (andychu) [inline library/func_misc.py:162]: This check looks good for now, though I also sketched this typed_args.Reader thing that's not implemented

Should discuss on Zulip

When we do proc argument binding, this discussion of the style will probably be more relevant

File: library/func_misc.py [hunk 3]
```diff
@@ -147,3 +147,44 @@ def Call(self, pos_args, named_args):
...
+            raise error.InvalidType("len() expects exactly one argument but %d were given" %
+                          n_args, loc.Missing)
+
+        x = pos_args[0]
+        UP_x = x
+
+        with tagswitch(x) as case:
+            if case(value_e.BashArray):
+                x = cast(value.BashArray, UP_x)
```
[the remark this answers is not in the document]
Comment (melvinw) [reply library/func_misc.py]: Oh, hmm... A few spec tests rely on `len()` being defined for `BashArray` and `BashAssoc`. Will patch those

**Comment by andychu** [comment]
Oh yeah sorry some of the tests like `var a = Dict()` are over-specified -- now I don't think we'll even have that synonym, it's just `var a = {}`

Remember a couple months ago I said the initial pass of YSH was "best first guess"

Looks like other tests are obsolete.  I did a lot of this pruning for the 0.16.0 release -- "breaking renames and YSH"

so yeah we should break these, let me make a pass right now

**Comment by andychu** [comment]
OK in doing the spec tests, I think we may want

- Bool.fromObj() -- equivalent of Python's bool()
- Int.fromStr(s, base) -- like Python `int('ff', 16)`
- Float.fromStr(s) -- like Python `float()`,  is this different than JSON notation for float?  Not sure
- `read --lines :branches` instead of `readarray -t :branches` (bash syntax)

Let me know what you think, looking for feedback

We kinda have to "scrub" the language for consistency / design while porting to C++

The basic idea is to do the MINIMAL thing that can express say a flag parser `argparse` or a test framework `describe`

Other stuff can be added later, but we want to avoid writing a lot of code, and avoid a big test matrix (hence getting rid of length of BashArray etc.)

**Comment by melvinw** [comment]
Ah great. Thanks for taking care of those tests

**Comment by melvinw** [comment]
> Oh yeah sorry some of the tests like var a = Dict() are over-specified -- now I don't think we'll even have that synonym, it's just var a = {}

Agreed. All `Dict()` provided in my mind was a way to copy, but I think we should provide a generic way to do that over all `value_t`. Something like `obj.Copy()` or `obj.Clone()`

> OK in doing the spec tests, I think we may want
> 
>     * Bool.fromObj() -- equivalent of Python's bool()

`Bool()` feels cleaner to me. Why do we want the dot syntax here?

>     * Int.fromStr(s, base) -- like Python `int('ff', 16)`

`fromStr()` makes more sense here. Do we want to support creating ints from floats too? e.g. python has
```
$ python
Python 2.7.18 (default, Jul  9 2023, 16:30:32)
[GCC 12.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> int(1.2)
1
```
>     * Float.fromStr(s) -- like Python `float()`,  is this different than JSON notation for float?  Not sure

Similar questions here? Couldn't this be `Float()` and it can accept both `value.Str` and `value.Int`?

**Pull request merged**